subs/pantry/src/Pantry/Archive.hs

```diff
@@ -387,41 +393,60 @@ parseArchive pli archive fp = do
             case Map.lookup (seSource se) blobs of
               Nothing -> error $ "Impossible: blob not found for: " ++ seSource se
               Just blobKey -> pure (sfp, TreeEntry blobKey (seType se))
-
           -- parse the cabal file and ensure it has the right name
-          (cabalPath, cabalEntry@(TreeEntry cabalBlobKey _)) <- findCabalFile pli tree
+          (cabalPath, cabalEntry@(TreeEntry cabalBlobKey _), btype) <- findCabalOrHpackFile pli tree
           mbs <- withStorage $ loadBlob cabalBlobKey
           bs <-
             case mbs of
               Nothing -> throwIO $ TreeReferencesMissingBlob pli cabalPath cabalBlobKey
               Just bs -> pure bs
-          (_warnings, gpd) <- rawParseGPD (Left pli) bs
+          cabalBs <- if btype == CabalFile
+                     then return bs
+                     else do
+                       getCabalBS bs
+          (_warnings, gpd) <- rawParseGPD (Left pli) cabalBs
           let ident@(PackageIdentifier name _) = package $ packageDescription gpd
-          when (cabalPath /= cabalFileName name) $
+          when (cabalPath /= cabalFileName name && btype == CabalFile) $
             throwIO $ WrongCabalFileName pli cabalPath name
 
           -- It's good! Store the tree, let's bounce
-          (_tid, treeKey) <- withStorage $ storeTree ident tree cabalEntry
+          (_tid, treeKey) <- withStorage $ storeTree ident tree cabalEntry btype
+          let tentry = if btype == CabalFile
```

Comment:
"t" here looks similar like the 1st letter in `treeKey`, not clear what it means and a bit confusing

Reply:
Renamed it. Hopefully the new variable `packageCabal` makes it clear.